packages/next-on-pages/templates/_worker.js/utils/fetch.ts

```diff
@@ -1,79 +1,121 @@
-				const resp = {
-					async arrayBuffer() {
-						return binaryContent;
-					},
-					get body(): ReadableStream<unknown> | null {
-						return new ReadableStream({
-							start(controller) {
-								return pump();
-								function pump() {
-									const b = Buffer.from(binaryContent);
-									controller.enqueue(b);
-									controller.close();
-									return pump();
-								}
-							},
-						});
-					},
-					async text() {
-						const b = Buffer.from(binaryContent);
-						return b.toString();
-					},
-					async json() {
-						const b = Buffer.from(binaryContent);
-						return JSON.stringify(b.toString());
-					},
-					async blob() {
-						return new Blob(binaryContent);
-					},
-				} as Response;
+		setRequestUserAgentIfNeeded(request);
+
+		return originalFetch(request);
+	};
+}
+
+/**
+ * This function checks if a given request is trying to fetch an inline if it is it returns a response containing a stream for the asset,
+ * otherwise returns null (signaling that the request hasn't been handled).
+ *
+ * This is necessary so that users can fetch urls such as: `new URL("file", import.meta.url)`
+ * (used for example with `@vercel/og`)
+ *
+ * Note: this function's aim is to mimic the following Next behavior:
+ * 	https://github.com/vercel/next.js/blob/6705c803021d3bdea7fec20e5d98f6899e49836d/packages/next/src/server/web/sandbox/fetch-inline-assets.ts
+ *
+ * @param request the request to handle
+ * @returns the response to return to the caller if the request was for an inline asset one (and the file exists), null otherwise
+ */
+async function handleInlineAssetRequest(request: Request) {
+	if (request.url.startsWith('blob:')) {
+		try {
+			const url = new URL(request.url);
+			const binaryContent = (
+				await import(`./__next-on-pages-dist__/assets/${url.pathname}.bin`)
+			).default;
 
-				// Note: clone is necessary so that body does work
-				resp.clone = (): Response => {
-					return resp;
-				};
+			// Note: we can't generate a real Response object here because this fetch might be called
+			//       at the top level of a dynamically imported module, and such cases produce the following
+			//       error:
+			//           Some functionality, such as asynchronous I/O, timeouts, and generating random values,
+			//           can only be performed while handling a request
+			//       this is a somewhat known workerd behavior (currently kept for security and performance reasons)
+			//
+			//       if the above issue/constraint were to change we should replace the following with a real Response object
+			const resp = {
+				async arrayBuffer() {
+					return binaryContent;
+				},
+				get body(): ReadableStream<unknown> | null {
+					return new ReadableStream({
+						start(controller) {
+							return pump();
+							function pump() {
+								const b = Buffer.from(binaryContent);
+								controller.enqueue(b);
+								controller.close();
+								return pump();
```

Comment:
I think this will cause a `Maximum call stack size exceeded` error

Since you have already enqueued all chunks and closed the controller, I don't think you need to recursively call pump

Reply:
ah ok I did tested it and it is actually happens to be working fine

when it calls `pump` the function runs twice once normally and the second time it must be throwing or something (even though it doesn't get surfaced) when it does `controller.enqueue` with the closed controller:
![Screenshot 2023-07-20 at 00 03 19](https://github.com/cloudflare/next-on-pages/assets/61631103/0652f1f3-8dbc-4ed6-8599-e9126e75c358)

![Screenshot 2023-07-20 at 00 03 48](https://github.com/cloudflare/next-on-pages/assets/61631103/f691c69c-d5d3-4ace-97e4-ed29e4dff4a4)

but yeah the code is wrong (and it doesn't need a pump function 🤦) I've updated it, thanks again for noticing it! ❤️  (it was a stupid copy-paste mistake on my part 😓)